### PR TITLE
Blog - How to Build GraphQL APIs Using Go

### PR DESCRIPTION
**Pull request by msali123**
Done with all checks.

### CODE REVIEW

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+ Title string `gorm:"not null"`
+ Content string `gorm:"not null"`
+ Author string `gorm:"not null; unique"`
+ Hero string `json:"Hero"`
+ Published_At string `json:"PublishedAt"`
+ Updated_At string `json:"UpdateAt"`
+}
+~~~
+
+Note: You can use the `Post` struct generated by gqlgen. However, it's good to take advantage of Gorm and describe your database structure with a few lines of code.
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:297]:
```suggestion
**Note**: You can use the `Post` struct generated by gqlgen. However, it's good to take advantage of Gorm and describe your database structure with a few lines of code.
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+ "github.com/jinzhu/gorm"
+ "go-graphql-api/dbmodels"
+)
+~~~
+
+You can still run the `go mod tidy` command to ensure any missing dependencies are added to your direct dependencies.
+
+Create the following variables:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:318]:
```suggestion
~~~{.go caption="mysql.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+// the db connection string
+var CONNECTION_STRING string = \
+"db_username:your_user_password@tcp(localhost:3306)/?charset=utf8&parseTime=True&loc=Local"
+~~~
+
+Your `CONNECTION_STRING` should reflect the credentials of your MySQL server, such as your user and the MySQL user password.
+
+Establish the connection to your database:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:332]:
```suggestion
~~~{.go caption="mysql.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+ // log all database operations performed by this connection
+ DBInstance.LogMode(true)
+}
+~~~
+
+The `CONNECTION_STRING` opens a connection to the MySQL server. The result of your connection will be logged in your terminal when running the application. `LogMode(true)` will allow the application to log all database operations to your terminal. You can set it to false if you don't need database operations output.
+
+Create a database:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:356]:
```suggestion
~~~{.go caption="mysql.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+ // make the database available to this connection
+ DBInstance.Exec("USE Blog_Posts")
+}
+~~~
+
+Instead of manually creating a database, use the Gorm ORM to handle this. It will execute the above query and create the database for you.
+
+Migrate Post Model to a database table:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:369]:
```suggestion
~~~{.go caption="mysql.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+
+~~~{.go caption="server.go"}
+import(
+  "go-graphql-api/database"
+)
+~~~
+
+Execute the following database functions:
+
+~~~{.bash caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:403]:
```suggestion
~~~{.go caption="server.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+database.ConnectDB()
+// create db
+database.CreateDB()
+// migrate the db with Post model
+database.MigrateDB()
+~~~
+
+Save the established database connection in the Resolver struct of your `srv` variable generated by gqlgen:
+
+~~~{.bash caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:414]:
```suggestion
~~~{.go caption="server.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+(generated.Config{Resolvers: &graph.Resolver{
+  Database: database.DBInstance,
+}}))
+~~~
+
+The database is now set to execute a connection and carry out the expected database operations.
+
+First, run the application to ensure the set database works as expected. To run the server, use [Air](https://github.com/cosmtrek/air) so that you can live reload the server when you make new changes. This way, focus on your code. Once you run the server once, Air will execute your new changes and reload the app for you. To install Air run:
+
+~~~{.go caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:425]:
```suggestion
~~~{.bash caption=">_"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+go install github.com/cosmtrek/air@latest
+~~~
+
+Then initialize it using:
+
+~~~{.go caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:431]:
```suggestion
~~~{.bash caption=">_"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+air init
+~~~
+
+Note: You may be required to delete the `tools.go` file created earlier. You can confidently delete this file as we no longer need it. Otherwise, Air will generate an error **tool.go:4:2: import "github.com/99designs/gqlgen" is a program, not an importable package.**
+
+Now that Air is set, at the root directory of your application, run:
+
+~~~{.go caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:439]:
```suggestion
~~~{.bash caption=">_"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 11]
```diff
@@ -0,0 +1,652 @@
...
+  Database: database.DBInstance,
+}}))
+~~~
+
+The database is now set to execute a connection and carry out the expected database operations.
+
+First, run the application to ensure the set database works as expected. To run the server, use [Air](https://github.com/cosmtrek/air) so that you can live reload the server when you make new changes. This way, focus on your code. Once you run the server once, Air will execute your new changes and reload the app for you. To install Air run:
+
+~~~{.go caption=">_"}
+go install github.com/cosmtrek/air@latest
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:426]:
```suggestion
$ go install github.com/cosmtrek/air@latest
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 12]
```diff
@@ -0,0 +1,652 @@
...
+~~~
+
+Then initialize it using:
+
+~~~{.go caption=">_"}
+air init
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:432]:
```suggestion
$ air init
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 13]
```diff
@@ -0,0 +1,652 @@
...
+~~~
+
+Note: You may be required to delete the `tools.go` file created earlier. You can confidently delete this file as we no longer need it. Otherwise, Air will generate an error **tool.go:4:2: import "github.com/99designs/gqlgen" is a program, not an importable package.**
+
+Now that Air is set, at the root directory of your application, run:
+
+~~~{.go caption=">_"}
+air
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:440]:
```suggestion
$ air
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+
+These mutations, however, have not yet been handled. They are still the boilerplate code. We need to modify them to handle the mutations logic. Head over to your `graph/schema.resolvers.go` and start working on the mutation methods as follows:
+
+#### Creating Posts Mutation
+
+First, add time to your imports. Some fields require time, and the time parameter format is added to the database.
+
+To create a post, modify the `CreatePost` method as follows:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:466]:
```suggestion
~~~{.go caption="graph/schema.resolvers.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+
+We can create values into the database using the SQL `Create()` method. It takes the parameter of the Post model that you want to add. The `Addpost` variable perfectly describes this.
+
+To execute the `CreatePost` mutation, ensure the server is still up and running. Otherwise, re-run your server using the `air` command. To test the API, open `http://localhost:8080/` to access your API GraphQL playground in the browser.
+
+The GraphQL playground is ready, and you see all the GraphQL root types for each kind of operation related to the created API.
+
+On your GraphQL playground, execute the following mutation:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:497]:
```suggestion
~~~{. caption=""}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 16]
```diff
@@ -0,0 +1,652 @@
...
+
+Go ahead and add different posts using the above schema as an example.
+
+You can confirm this addition on your MySQL database to see if a new post got added.
+
+#### Updating Posts Mutation
+
+Modify the `UpdatePost()` method to add the update mutation resolver:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:526]:
```suggestion
~~~{.go caption="graph/schema.resolvers.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 17]
```diff
@@ -0,0 +1,652 @@
...
+ Updatepost.ID = postID
+ return &Updatepost, nil
+}
+~~~
+
+First, add the fields to execute on the update mutation as described in the `Updatepost` variable. To alter any available posts, you need to execute a SQL query matching the updated post. The `Where` clause takes the post id parameter to check the mutated database record.
+
+Updating mutates your saved data. Therefore, a mutation of new data is sent to your database to update a post. To edit an existing post, use the following schema on your GraphQL playground:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:551]:
```suggestion
~~~{. caption=""}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 18]
```diff
@@ -0,0 +1,652 @@
...
+
+Go ahead and check if the changes were implemented to the selected post.
+
+### Building Go Query Resolver
+
+#### Retrieve All Posts Query
+
+Add the following modifications to the `GetAllPosts` method in `graph/schema.resolvers.go` to add a resolver that fetches all posts.
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:575]:
```suggestion
~~~{.go caption="graph/schema.resolvers.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 19]
```diff
@@ -0,0 +1,652 @@
...
+ }
+ return posts, nil
+}
+~~~
+
+Create a variable `post` to save all the fields you want to get from the post model. This example fetches all fields. The database will execute the models that match the posts table using the SQL `Find()` method to find records that match the given conditions.
+
+Send the following query to retrieve all the posts saved in your database:
+
+~~~{.bash caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:595]:
```suggestion
~~~{. caption=""}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 20]
```diff
@@ -0,0 +1,652 @@
...
+    Updated_At
+  }
+}
+~~~
+
+#### Retrieve A Single Post Query
+
+Modify the `GetOnePost` method to execute the query resolver for getting a single post.
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:613]:
```suggestion
~~~{.go caption="graph/schema.resolvers.go"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 21]
```diff
@@ -0,0 +1,652 @@
...
+ 
+ return &post, nil
+}
+~~~
+
+Just like `GetAllPosts`, use the `Find()` method to find records that match the given conditions. In this case, `Find()` will fulfill the condition of getting one post. Therefore, the id parameter will be executed to fetch the post that matches the passed id.
+
+To get a single post, pass the post id to your query as follows:
+
+~~~{.js caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:632]:
```suggestion
~~~{. caption=""}
```

**Review by jalletto — CHANGES_REQUESTED** [review]
Hey Sameer, 

Made some changes to the code examples. In this case, the author included the code in a github repo https://github.com/Rose-stack/go-mysql-graphql-api. So in cases like this in the future, you can look to the source code to be sure you are getting the correct language and file.

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 22]
```diff
@@ -0,0 +1,652 @@
...
+- Implementing the Resolvers
+  - Building Go Query Resolver
+  - Building Go Mutation Resolver
+- Testing the GraphQL API
+  - Testing Query method
+  - Testing Mutation method
+- Conclusion
+
+The purpose of an API is to communicate between the client and the server. There are different protocols and architectures that you can use to create API services and consume resources. This guide will help you learn how to implement a GraphQL API using Golang and consume data from a server. You will use the [MySQL](/blog/docker-mysql) database server to perform different operations using the GraphQL API.
+
```
[the remark this answers is not in the document]
Comment (msali123) [reply blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:38]: Done

**Comment by msali123** [comment]
@jalletto sorry for too much code area mistakes.

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 23]
```diff
@@ -0,0 +1,628 @@
...
+Initialize gqlgen to build your boilerplate Go GraphQL API:
+
+~~~{.go caption=">_"}
+go run github.com/99designs/gqlgen init
+~~~
+
+This will create a basic GraphQL API with the following file structure:
+
+~~~{. caption=""}
```
[the remark this answers is not in the document]
Comment (msali123) [reply blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:74]: Josh do I need to change only code blocks for output or text?

Or for the code blocks as well like this?
~~~{.go caption=">_"}

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 24]
```diff
@@ -0,0 +1,629 @@
...
+
+## Gqlgen for Go GraphQL APIs
+
+Gqlgen is a Go GraphQL library that allows you to build robust GraphQL servers without creating everything from scratch. Gqlgen is a schema first library meaning it creates the API schema using GraphQL [schema definition language](https://graphql.org/learn/schema/). Using that schema as the input, gqlgen generates the GraphQL server code. This way, you can build applications quickly. All you need is to implement the core logic of your GraphQL resolvers.
+
+## Generating Boilerplate Go GraphQL API Using Gqlgen
+
+Let's dive in and implement the API using gqlgen. First, initialize a [Golang](/blog/top-3-resources-to-learn-golang-in-2021) application inside the directory where you want the project to live:
+
+~~~{.go caption=">_"}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:40]:
```suggestion
~~~{.bash caption=">_"}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 25]
```diff
@@ -0,0 +1,629 @@
...
+
+We can create values into the database using the SQL `Create()` method. It takes the parameter of the Post model that you want to add. The `Addpost` variable perfectly describes this.
+
+To execute the `CreatePost` mutation, ensure the server is still up and running. Otherwise, re-run your server using the `air` command. To test the API, open `http://localhost:8080/` to access your API GraphQL playground in the browser.
+
+The GraphQL playground is ready, and you see all the GraphQL root types for each kind of operation related to the created API.
+
+On your GraphQL playground, execute the following mutation:
+
+~~~{.bash caption=""}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:473]:
```suggestion
~~~{.text caption=""}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 26]
```diff
@@ -0,0 +1,629 @@
...
+ Updatepost.ID = postID
+  return &Updatepost, nil
+}
+~~~
+
+First, add the fields to execute on the update mutation as described in the `Updatepost` variable. To alter any available posts, you need to execute a SQL query matching the updated post. The `Where` clause takes the post id parameter to check the mutated database record.
+
+Updating mutates your saved data. Therefore, a mutation of new data is sent to your database to update a post. To edit an existing post, use the following schema on your GraphQL playground:
+
+~~~{.bash caption=""}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:527]:
```suggestion
~~~{.text caption=""}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 27]
```diff
@@ -0,0 +1,629 @@
...
+ }
+ return posts, nil
+}
+~~~
+
+Create a variable `post` to save all the fields you want to get from the post model. This example fetches all fields. The database will execute the models that match the posts table using the SQL `Find()` method to find records that match the given conditions.
+
+Send the following query to retrieve all the posts saved in your database:
+
+~~~{.bash caption=""}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:571]:
```suggestion
~~~{.text caption=""}
```

File: blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md [hunk 28]
```diff
@@ -0,0 +1,629 @@
...
+ 
+ return &post, nil
+}
+~~~
+
+Just like `GetAllPosts`, use the `Find()` method to find records that match the given conditions. In this case, `Find()` will fulfill the condition of getting one post. Therefore, the id parameter will be executed to fetch the post that matches the passed id.
+
+To get a single post, pass the post id to your query as follows:
+
+~~~{.bash caption=""}
```
Comment (jalletto) [inline blog/_posts/2022-10-19-build-GraphQL-APIs-Go.md:609]:
```suggestion
~~~{.text caption=""}
```

**Review by jalletto — CHANGES_REQUESTED** [review]

**Review by jalletto — APPROVED** [review]

**Pull request merged**